Membership and constants

jamalex opened the pull request:
## Summary

### Primary change made in this PR

- Introduced the `Membership` concept/model, which is similar to a `Role`, in that it links a `FacilityUser` and a `Collection`, but different in a fundamental way (sort of the inverse of a `Role`):
  - Having a `Role` for a `Collection` means also having the same type of role for all **descendants** of that collection. A `Role` gives *you* certain permissions with respect to collections or their contained members. We currently have two kinds of roles: "admin" and "coach".
  - Having a `Membership` in a `Collection` means having the same type of membership in all **ancestors** of that collection. A `Membership` can give *others* permissions with respect to *you*. There is only one kind of membership: "learner".

@MCGallaspy, this change above is what I wanted to chat with you about, to get quick feedback, before I proceed. Can we chat for 20 mins tomorrow morning?

### Other changes made in the PR

- Separated constants out into separate files, as recommended [here](http://stackoverflow.com/questions/11111632/python-best-cleanest-way-to-define-constant-lists-or-dictionarys)
- Relativized some imports.

## TODO

- [x] Have tests been written for the new code?
- [ ] Has documentation been written/updated?

jamalex commented on kolibri/auth/constants/relation_kinds.py:
```diff
@@ -0,0 +1,12 @@
+"""
```
Don't mind this one for now; it will come into play in the next PR.

jamalex replied:
(In fact, it was deleted in the next PR)

MCGallaspy commented:
Do you want me to review and merge this now? Or are you planning on adding more to this PR?

jamalex commented:
Sorry, was in meetings, just seeing this. Can be reviewed/merged, as it will make the next PR cleaner.

MCGallaspy commented on kolibri/auth/models.py:
```diff
@@ -242,14 +233,44 @@ def remove_user(self, user, role_kind):
+        assert isinstance(user, FacilityUser), "Only FacilityUsers can be removed from a collection."
+
+        # delete the appropriate membership, if it exists
+        Membership.objects.filter(user=user, collection=self).delete()
```
For this and `remove_user` above I think it's appropriate to either return the value of this LOC, or use `get` instead of `filter`. This method as written will quietly do nothing if no objects are found. This can be ascertained by looking at `delete`'s return value, or by using `get` which will raise an exception if no object is found.

jamalex replied:
Good call, and very interesting -- I had never inspected the return value of `delete` before; neat how it also tells you how many related models of each type were deleted.

Updating this to return the results.

jamalex replied:
Actually, since there shouldn't ever be more than one deleted, and no related models should be deleted, it'll be cleaner to just return True/False to indicate whether anything was deleted.

MCGallaspy commented on kolibri/auth/test/test_collection_and_role_models.py:
```diff
@@ -84,9 +89,9 @@ def test_delete_facility_pt3(self):
 
     def test_delete_facility_user(self):
         """ Deleting a FacilityUser should delete associated Roles """
```
Missed a docstring change.

jamalex commented:
All comments have now been addressed. I'll rebase #51 to the latest on this branch.

MCGallaspy commented on kolibri/auth/test/test_collection_and_role_models.py:
```diff
@@ -1,56 +1,72 @@
+
+    def test_remove_indirect_admin_role(self):
+        """ Trying to remove the admin role for a a Facility admin from a descendent classroom doesn't actually remove anything. """
+        self.assertFalse(self.cr.remove_admin(self.facility_admin))
```
Seems like there's a distinction to be made between failing to remove a role because it doesn't exist vs failing to remove a role because somehow the request to do so doesn't make sense. Consider the following snippets, which might appear in a view for user management:

``` python
if classroom.remove_admin(facility_admin):
    # The role was successfully removed
    return SuccessResponse("Role removed!")
else:
    return FailureResponse("Unable to remove role!")  # No indication as to why
```

vs

``` python
try:
    removed = classroom.remove_admin(facility_admin)
except KolibriCoolHierarchyError:
    return FailureResponse("Can't remove, user possesses role from parent Facility!")
else:
    if removed:
        return SuccessResponse("Role removed!")
    else:
        return FailureResponse("User does not possess that role!")
```

Is it useful to make such a distinction in your opinion?

jamalex replied:
Could be useful. I think if we're going to throw exceptions, it's better to just throw exceptions for both cases (user not having role at all, and user having the role via the collection hierarchy). I'll add something in for that.

jamalex commented:
Responded to comment:
https://github.com/learningequality/kolibri/pull/50#discussion_r57761906

And added a commit to address this (and add a couple other similar exception types to make things nicer to handle):
https://github.com/learningequality/kolibri/pull/50/commits/04553bffd21dc76563ea95ff851b932ec3c2dcc8

MCGallaspy commented:
Plan to address test failures?

MCGallaspy commented:
By the way, I think you had mentioned that you plan on writing documentation later, perhaps in #51. Is that the case, or should I expect some here?

jamalex commented:
> Plan to address test failures?

Failures are due to TDD; in order to properly determine (as requested) what type of exception to throw within the role management methods, they now reference methods (`has_role_for` and `is_member`) that won't be defined until #51.

jamalex commented:
> By the way, I think you had mentioned that you plan on writing documentation later, perhaps in #51. Is that the case, or should I expect some here?

Documentation overhaul will come in #51, after initial review is complete, and the plan for the interface is solidified there.

MCGallaspy commented:
Unfortunately merging would then cause other's branches based off of master to fail. So I guess we have a few options:

1. Be okay with that.
2. Target a new feature branch instead of master and be okay with failures there.
3. Close this PR and favor #51 instead. (But then it becomes a pretty big PR.)
4. Implement the functionality here for tests to pass.

I favor 4 first and then 2 -- what do you think?

jamalex commented:
Sure, I'll implement the missing methods in this PR, and then can rebase #51 against this one again.

jamalex commented:
Was dev'ing with Python 2 -- a couple test failures with Python 3; fixing now.

jamalex commented on kolibri/auth/test/test_collections_and_roles.py:
```diff
@@ -1,270 +0,0 @@
-from __future__ import absolute_import, print_function, unicode_literals
```
Renamed and reworked in test_collection_and_role_models.py, above.

jamalex commented:
Note: Main implementation of role-checking logic is in here (as requested, to get tests passing), but most usage happens in #51.

MCGallaspy commented on kolibri/auth/models.py:
```diff
@@ -140,6 +140,60 @@ def get_full_name(self):
 
+    def is_member_of(self, coll):
+        """
+        Determine whether this user has the specified role kind in relation to the target user.
```
Is this docstring inaccurate?

jamalex replied:
Yes! Thanks, will fix.

jamalex replied:
Fixed.

MCGallaspy commented on kolibri/auth/filters.py:
```diff
@@ -0,0 +1,185 @@
+from .constants import collection_kinds
+
+
+class HierarchyRelationsFilter(object):
```
This is a complex class that seems to deal directly with SQL statements. I think the implication is that we might create something that is not transferable to other database backends outside of sqlite -- is that true? It might not matter if we don't intend for Kolibri to actually be used with other database backends. But if we _do_ then we should try to answer this question definitively, by setting up a test case that uses the target database backend. What do you think?

Secondly, I think this class in particular will benefit from comprehensive test coverage, but on this branch the test coverage for this class is not comprehensive, and indeed in `models.py` some usages of this class are missed. The coverage is also incomplete on #51. As a tactical review choice, I would substitute 100% test coverage for line-by-line scrutiny. Can you get 100% coverage in #51? Otherwise I would feel obliged to spend more time on this PR.

jamalex replied:
> I think the implication is that we might create something that is not transferable to other database backends outside of sqlite -- is that true?

There's no flavor-specific SQL in there (i.e. not tied to sqlite) -- just standard joins and WHERE conditions. We do want to use Kolibri with other DB backends (like Postgres), but I don't think this code in particular is more vulnerable to DB differences than standard Django code (e.g. we've seen issues in KA Lite of different DB level integrity enforcement (such as uniqueness) in sqlite vs mysql. So, I think we'll want to (occasionally) run the full test suite using these alternative database engines, once we start using them, but don't think it makes sense to try to integrate them into the test suite itself (would probably make for much slower tests).

> I think this class in particular will benefit from comprehensive test coverage ... Can you get 100% coverage in #51? Otherwise I would feel obliged to spend more time on this PR.

Agreed. I will aim to get 100% coverage in #51, and delete anything that I'm not able to touch there.

MCGallaspy replied:
> So, I think we'll want to (occasionally) run the full test suite using these alternative database engines, once we start using them, but don't think it makes sense to try to integrate them into the test suite itself (would probably make for much slower tests).

I agree that it could make for slower tests -- how _much_ slower probably depends on a lot of factors. For instance, we may only wish to run a judiciously chosen _subset_ of the full test suite with postgres. Or we could always carefully identify bottlenecks using a postgres test db and optimize for them e.g. by preserving the test database in between runs (avoiding the need to rebuild it every time). In fact, Django has some facilities for this already.

Moreover for me the takeaway from the case you mention regarding KA Lite is to test early and often with the target stack -- if you plan to use postgres as a backend _eventually_, IMO it would save time ultimately to start testing with it _now_. Otherwise we risk quietly building up a lot of changes that we'll need to slog through when we _do_ finally want to use it. Bear in mind that Django has a lot of scrutiny and provides an ORM that's meant to provide a consistent interface to any database backend, yet _still_ has differences between backends. In contrast this project has orders of magnitude less scrutiny, and this code in particular is presumably being scrutinized only by you and me.

Personally, because I'm not an expert with postgres and I'm not able to clearly identify what will be problematic and what will not, I would hedge my bets by testing early, often, and automatically as opposed to later, occasionally, and manually. That's my 2 cents. If you don't find my reasoning compelling enough, I'll merge this.

jamalex replied:
I agree it would be good to have selective tests run with postgres -- I'm just not familiar enough with CircleCI to know how hard this is to set up (obviously we don't want every user to need to have postgres installed locally in order to run tests, but I imagine there's some mechanism for spinning up a postgres database within CircleCI, and then we can have those tests run selectively there). If you think that's easy to set up, let's do it (I'd need direction). Otherwise, I'd say let's cross that bridge when we get there.

MCGallaspy replied:
> If you think that's easy to set up, let's do it (I'd need direction). Otherwise, I'd say let's cross that bridge when we get there.

I do think it would be fairly straightforward to set up:
1. Get postgres running in our travis environment: https://docs.travis-ci.com/user/database-setup/#PostgreSQL
2. Create a new settings module that uses a postgres test database: https://docs.djangoproject.com/en/1.9/topics/testing/overview/#the-test-database
3. Add a new tox configuration to run the tests using that settings module.

If supporting postgres is important to you, then we're at the bridge IMO. So I'm going to set it up right now.

MCGallaspy replied:
I'm merging this now and I'll work on the test config separately.

jamalex replied:
Thanks.

jamalex commented:
Addressed the comments here, switching focus to coverage and documentation in #51. If satisfactory, this PR can now be merged, and focus shifted to #51.
